field_audit/auditors.py

```diff
@@ -0,0 +1,130 @@
+    (and in what order) ``BaseAuditor`` subclass instances are used to acquire
+    "changed by" information.
+
+    The auditor chain can be customized by defining a list of 'BaseAuditor'
+    subclass paths via the ``FIELD_AUDIT_AUDITORS`` settings attribute.
+    """
+
+    def setup_auditors(self):
+        """Populate the auditors chain, possibly defined in settings.
+
+        This method is called at app ready time, and must be called before the
+        ``dispatch()`` method can be used.
+        """
+        from django.conf import settings
+        auditors_attr = "FIELD_AUDIT_AUDITORS"
+        if hasattr(settings, auditors_attr):
+            self.auditors = []
+            for auditor_path in getattr(settings, auditors_attr):
+                auditor_class = class_import_helper(
+                    auditor_path,
+                    f"{auditors_attr!r} item",
+                    BaseAuditor,
+                )
+                self.auditors.append(auditor_class())
+        else:
+            self.auditors = [RequestAuditor(), SystemUserAuditor()]
+
+    def dispatch(self, request):
+        """Cycles through the auditors chain and returns the first non-None
+        value returned by a call to ``auditor.changed_by(request)``.
+
+        :param request: Django request to be audited (or ``None``).
+        :returns: JSON-serializable value (or ``None`` if chain is exhausted).
+        """
+        for auditor in self.auditors:
+            changed_by = auditor.changed_by(request)
+            if changed_by is not None:
+                return changed_by
+        return None
+
+
+audit_dispatcher = _AuditDispatcher()
+
+
+class BaseAuditor:
+    """Abstract class for the Auditor API. Subclasses are used to return
+    "changed by" information associated with an event which they know how to
+    audit.
+
+    BaseAuditor subclasses must define the following:
+    - ``changed_by()`` method that returns a JSON-serializable object of
+      information for events it knows how to audit (or ``None`` otherwise).
+    """
+
+    def changed_by(self, request):
+        raise NotImplementedError("subclasses must implement 'changed_by()'")
+
+
+class RequestAuditor(BaseAuditor):
+    """Auditor class for getting users from authenticated requests."""
+
+    def changed_by(self, request):
+        if not request:
+            # cannot provide a request user without a request
+            return None
+        if request.user.is_authenticated:
+            return {
+                "user_type": USER_TYPE_REQUEST,
+                "username": request.user.username,
+            }
+        return None
+
+
+class SystemUserAuditor(BaseAuditor):
+    """Auditor class for getting OS usernames."""
+
+    def __init__(self):
+        self.has_who_bin = True
+
+    def changed_by(self, request):
+        if request:
+            # never return a system user when a request is provided
+            return None
```

Comment:
The problem is that the current implementation of the auditor chain "behavior" doesn't provide a mechanism for auditor objects _in_ the chain to be aware of each other. As such, if a "request auditor" (one that knows how to audit an HTTP request) isn't able to extract audit information about a specific request, it just returns `None`: ignorant to whether it's the _last_ "request auditor" or if there's another (more specialized one) that will know what to do with the request.

With that context, in the current implementation using the default audit chain (e.g. `[RequestAuditor(), SystemUserAuditor()]`), if an unauthenticated request is run down the chain and the `SystemUserAuditor` does _not_ return `None`, the resulting `changed_by` value will report the Django WSGI server's OS user, which is probably not an an "expected value" in that scenario.

This does feel a bit like I'm "guessing in the face of ambiguity".  Perhaps it would be better to implement another auditor in the chain (something like `UnauthenticatedRequestAuditor`) that returns a static value (like `{"user_type": None}`) and document why you should consider using it you define your own auditor chain and wish to short-circuit on unauthenticated web requests? In this case it might be better to also implement a constant "I don't know" value (e.g. `BaseAuditor.NOT_FOUND`) so an auditor in the chain can return `None` meaning "I know the `changed_by` value is actually `None`.

Reply:
Offlined with @millerdev about this. I will make an update with a cleaner solution.